doc/neps/nep-0035-array-creation-dispatch-with-array-function.rst

```diff
@@ -8,16 +8,176 @@ NEP 35 — Array Creation Dispatching With __array_function__
+and ensures correctness with ``__array_function__``. Note that Dask can perform
+computations on different sorts of objects, like dataframes, bags and arrays,
+here we will focus strictly on arrays, which are the objects we can use
+``__array_function__`` with.
+
+Dask uses a graph computing model, meaning it breaks down a large problem in
+many smaller problems and merges their results to reach the final result. To
+break the problem down into smaller ones, Dask also breaks arrays into smaller
+arrays that it calls "chunks". A Dask array can thus consist of one or more
+chunks and they may be of different types. However, in the context of
+``__array_function__``, Dask only allows chunks of the same type; for example,
+a Dask array can be formed of several NumPy arrays or several CuPy arrays, but
+not a mix of both.
+
+To avoid mismatched types during computation, Dask keeps an attribute ``_meta`` as
+part of its array throughout computation: this attribute is used to both predict
+the output type at graph creation time, and to create any intermediary arrays
+that are necessary within some function's computation. Going back to our
+previous example, we can use ``_meta`` information to identify what kind of
+array we would use for padding, as seen below:
+
+.. code:: python
+
+    import numpy as np
+    import cupy
+    import dask.array as da
+    from dask.array.utils import meta_from_array
+
+    def my_dask_pad(arr, padding):
+        padding = np.array(padding, like=meta_from_array(arr))
+        return np.concatenate((padding, arr, padding))
+
+    # Returns dask.array<concatenate, shape=(9,), dtype=int64, chunksize=(5,), chunktype=numpy.ndarray>
+    my_dask_pad(da.arange(5), [-1, -1])
+
+    # Returns dask.array<concatenate, shape=(9,), dtype=int64, chunksize=(5,), chunktype=cupy.ndarray>
+    my_dask_pad(da.from_array(cupy.arange(5)), [-1, -1])
+
+Note how ``chunktype`` in the return value above changes from
+``numpy.ndarray`` in the first ``my_dask_pad`` call to ``cupy.ndarray`` in the
+second. We have also renamed the function to ``my_dask_pad`` in this example
+with the intent to make it clear that this is how Dask would implement such
+functionality, should it need to do so, as it requires Dask's internal tools
+that are not of much use elsewhere.
+
+To enable proper identification of the array type we use Dask's utility function
+``meta_from_array``, which was introduced as part of the work to support
+``__array_function__``, allowing Dask to handle ``_meta`` appropriately. That
+function is primarily targeted at the library's internal usage to ensure chunks
+are created with correct types. Without the ``like=`` argument, it would be
+impossible to ensure ``my_pad`` creates a padding array with a type matching
+that of the input array, which would cause a ``TypeError`` exception to
+be raised by CuPy, as discussed above would happen to the CuPy case alone.
```

Comment:
So to be clear:

* `my_pad(dask_array)` should work just fine, thanks to the `like` argument and `__array_function__`
*  `my_dask_pad(dask_array)` is an example of what code internal to dask might look like.

Mentioning `my_pad`  without comparing it to `my_dask_pad` makes reading this a little confusing to me

At any rate, it seems that `meta_from_array` is redundant here. Perhaps a more convincing example would be something like
```python
# disclaimer: I have never used dasked, I am guessing the API from your summary!
def pad_chunks(arr: dask.array, padding ):
    # we could use `np.array(padding, like=arr).chunks[0]`, but this is more direct
    padding = np.array(padding, like=meta_from_array(arr))
    return dask,from_chunks([padding] + arr.chunks + [padding])
```
where the key point is you don't actually need `meta_from_array` except:
* Inside `__array_function__`
* As an optimization inside dask-specific functions.

Reply:
> my_pad(dask_array) should work just fine, thanks to the like argument and __array_function__

It would work fine for a "pure" Dask array, that's backed by NumPy. It won't work fine for a CuPy-backed Dask array, which needs to be handled accordingly, thus the `meta_from_array` call.

`meta_from_array` would be redundant if a Dask Array wasn't and array wrapping another type of array, which is why we need to handle it properly.

Anyway, I tried improving the text in https://github.com/numpy/numpy/pull/17093/commits/57f78df125892eaa4539913759f3ac29ff4aef6a , can you check if it makes things clearer?